simulator/main.rs

```diff
@@ -67,6 +116,343 @@ enum Value {
     Blob(Vec<u8>),
 }
 
+impl ArbitraryFrom<Vec<&Value>> for Value {
+    fn arbitrary_from<R: Rng>(rng: &mut R, t: &Vec<&Value>) -> Self {
+        if t.is_empty() {
+            return Value::Null;
+        }
+
+        let index = rng.gen_range(0..t.len());
+        t[index].clone()
+    }
+}
+
+impl ArbitraryFrom<ColumnType> for Value {
+    fn arbitrary_from<R: Rng>(rng: &mut R, t: &ColumnType) -> Self {
+        match t {
+            ColumnType::Integer => Value::Integer(rng.gen_range(i64::MIN..i64::MAX)),
+            ColumnType::Float => Value::Float(rng.gen_range(-1e10..1e10)),
+            ColumnType::Text => Value::Text(gen_random_text(rng)),
+            ColumnType::Blob => Value::Blob(gen_random_text(rng).as_bytes().to_vec()),
+        }
+    }
+}
+
+struct LTValue(Value);
+
+impl ArbitraryFrom<Vec<&Value>> for LTValue {
+    fn arbitrary_from<R: Rng>(rng: &mut R, t: &Vec<&Value>) -> Self {
+        if t.is_empty() {
+            return LTValue(Value::Null);
+        }
+
+        let index = rng.gen_range(0..t.len());
+        LTValue::arbitrary_from(rng, t[index])
+    }
+}
+
+impl ArbitraryFrom<Value> for LTValue {
+    fn arbitrary_from<R: Rng>(rng: &mut R, t: &Value) -> Self {
+        match t {
+            Value::Integer(i) => LTValue(Value::Integer(rng.gen_range(i64::MIN..*i - 1))),
+            Value::Float(f) => LTValue(Value::Float(rng.gen_range(-1e10..*f - 1.0))),
+            Value::Text(t) => {
+                // Either shorten the string, or make at least one character smaller and mutate the rest
+                let mut t = t.clone();
+                if rng.gen_bool(0.01) {
+                    t.pop();
+                    LTValue(Value::Text(t))
+                } else {
+                    let index = rng.gen_range(0..t.len());
+                    let mut t = t.into_bytes();
+                    t[index] -= 1;
+                    // Mutate the rest of the string
+                    for i in (index + 1)..t.len() {
+                        t[i] = rng.gen_range(0..=255);
+                    }
+                    LTValue(Value::Text(String::from_utf8(t).unwrap()))
```

Comment:
What is happening here

Reply:
There are two ways of generating a smaller text with respect to the lexicographic ordering:

1- `abc` is "smaller" than `abcd`, in the sense that if two strings share the same prefix, shorter is less than the longer. In 1% of the cases, the code removes the last element of the given value to create a shorter one.

2- `abc` is "smaller" than `aac`, or `aaz`, as the lexicographic ordering is the comparison of two char arrays starting with the leftmost character. The code first picks an index to `make smaller`, as in converting `abc` to `aac` by mutating `b` to `a`, and than randomly generating the right side of the picked index, as it is not important anymore.